cocotb/_compat.py

```diff
@@ -0,0 +1,109 @@
+"""
+Backports and compatibility shims for newer python features.
+
+These are for internal use - users should use a third party library like `six`
+if they want to use these shims in their own code
+"""
+import sys
+
+# This is six.integer_types
+if sys.version_info.major >= 3:
+    integer_types = (int,)
+else:
+    integer_types = (int, long)  # noqa
+
+
+# This is essentially six.exec_
+if sys.version_info.major == 3:
+    # this has to not be a syntax error in py2
+    import builtins
+    exec_ = getattr(builtins, 'exec')
+else:
+    # this has to not be a syntax error in py3
+    def exec_(_code_, _globs_=None, _locs_=None):
+        """Execute code in a namespace."""
+        if _globs_ is None:
+            frame = sys._getframe(1)
+            _globs_ = frame.f_globals
+            if _locs_ is None:
+                _locs_ = frame.f_locals
+            del frame
+        elif _locs_ is None:
+            _locs_ = _globs_
+        exec("""exec _code_ in _globs_, _locs_""")
+
+
+# this is six.with_metaclass, with a clearer docstring
+def with_metaclass(meta, *bases):
+    """This provides:
+
+    .. code-block:: python
+
+        class Foo(with_metaclass(Meta, Base1, Base2)): pass
+
+    which is a unifying syntax for:
+
+    .. code-block:: python
+
+        # Python 3
+        class Foo(Base1, Base2, metaclass=Meta): pass
+
+        # Python 2
+        class Foo(Base1, Base2):
+            __metaclass__ = Meta
+    """
+    # This requires a bit of explanation: the basic idea is to make a dummy
+    # metaclass for one level of class instantiation that replaces itself with
+    # the actual metaclass.
+    class metaclass(type):
+
+        def __new__(cls, name, this_bases, d):
+            return meta(name, bases, d)
+
+        @classmethod
+        def __prepare__(cls, name, this_bases):
+            return meta.__prepare__(name, bases)
+    return type.__new__(metaclass, 'temporary_class', (), {})
+
+
+# this is six.raise_from
+if sys.version_info[:2] == (3, 2):
+    exec_("""def raise_from(value, from_value):
+    try:
+        if from_value is None:
+            raise value
+        raise value from from_value
+    finally:
+        value = None
+    """)
```

Comment:
We don't support Python 3.2 any more

Reply:
Sure, that's fine.

In general, our policy has been "all officially supported Python versions". I thought we had documented that somewhere more prominently, but I didn't find it except for the latest release notes "Python 2.6 is officially not supported any more. cocotb supports Python 2.7 and Python 3.5+." (https://cocotb.readthedocs.io/en/latest/release_notes.html)